### PR TITLE
chore(DTFS2-6892): commonise london timezone

### PR DESCRIPTION
**Pull request by AlexBramhill**
## Introduction :pencil2:
As part of sso, I will need to use the london iana timezone. It is not currently commonised.

## Resolution :heavy_check_mark:
This ticket seeks to commonise the london timezone (outside of tests which are referring to it as a string), as well as creating a IanaTimezone type.

## Miscellaneous :heavy_plus_sign:
- The `IanaTimezone` type is broad -- being a `string`. This is because calling the `Intl.supportedValuesOf('timeZone')` function returns a broad type (`string`). I do not want to overly specify this type by making a static list of iana timezones in this PR.
- I've destructured imports as per the existing convention
- I've tidied up mismatching `require` and `import` on some js files

### CODE REVIEW

File: portal/server/nunjucks-configuration/filter-localiseTimestamp.ts [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import { IanaTimezone } from '@ukef/dtfs2-common';
```
Comment (AlexBramhill) [inline portal/server/nunjucks-configuration/filter-localiseTimestamp.ts:1]: commonising this file with the other filter file is not in scope for this ticket, as this is a large epic (sso)

File: portal/component-tests/admin/user-edit.component-test.js [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
+} = require('@ukef/dtfs2-common');
+const pageRenderer = require('../pageRenderer');
 
 const page = 'admin/user-edit.njk';
 const render = pageRenderer(page);
 
 describe(page, () => {
   const adminUser = {
-    timezone: LONDON_TIMEZONE,
+    timezone: LONDON,
```
Comment (ttbarnes) [inline portal/component-tests/admin/user-edit.component-test.js:11]: Some files now `TIMEZONES.LONDON`, others `LONDON`, can we be consistent please? I think either works

File: portal/component-tests/contract/components/contract-overview-table.component-test.js [hunk 3]
```diff
@@ -78,11 +81,7 @@ describe(component, () => {
       wrapper
         .expectText('[data-cy="submissionDate"]')
         .toRead(
-          formatInTimeZone(
-            new Date(dealWithManualInclusionApplicationSubmissionDate.details.manualInclusionApplicationSubmissionDate),
-            LONDON_TIMEZONE,
-            FULL_DATE,
-          ),
+          formatInTimeZone(new Date(dealWithManualInclusionApplicationSubmissionDate.details.manualInclusionApplicationSubmissionDate), LONDON, FULL_DATE),
```
Comment (ttbarnes) [inline portal/component-tests/contract/components/contract-overview-table.component-test.js:84]: Whilst we're in here, any chance we can destructure the date so this reads nicer? 

```js
formatInTimeZone(new Date(manualInclusionApplicationSubmissionDate), LONDON, FULL_DATE),
```

We could also rename `dealWithManualInclusionApplicationSubmissionDate` to `dealWithMiaSubmissionDate`

Appreciate this isn't the goal of the PR but whilst we're here, would be great

File: libs/common/src/constants/date-timezones.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { IanaTimezone } from '../types';
+
+/**
+ * IANA timezones used in the codebase
+ * Note - If needing a full list, we should look to Intl.supportedValuesOf('timeZone')
+ * {@link https://www.iana.org/time-zones}
+ */
+export const IANA_TIMEZONES = { LONDON: 'Europe/London' } as const satisfies Record<string, IanaTimezone>;
```
Comment (ttbarnes) [inline libs/common/src/constants/date-timezones.ts:8]: Can we match the file name with the const please

File: portal/server/nunjucks-configuration/filter-localiseTimestamp.test.ts [hunk 5]
```diff
@@ -1,6 +1,8 @@
-import { LONDON_TIMEZONE } from '../constants/date';
+import { IANA_TIMEZONES } from '@ukef/dtfs2-common';
```
Comment (ttbarnes) [inline portal/server/nunjucks-configuration/filter-localiseTimestamp.test.ts:1]: Some are importing `TIMEZONES`, others `IANA_TIMEZONES`, can they be consistent?

**Review by ttbarnes — CHANGES_REQUESTED** [review]

File: libs/common/src/constants/date-timezones.ts (hunk 4 above)
Comment (AlexBramhill) [reply libs/common/src/constants/date-timezones.ts:8]: updated -- kept date prefix for ease of finding in file browser

File: portal/component-tests/admin/user-edit.component-test.js (hunk 2 above)
Comment (AlexBramhill) [reply portal/component-tests/admin/user-edit.component-test.js:11]: updated

**Review by ttbarnes — APPROVED** [review]

File: libs/common/src/constants/date-iana-timezones.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { DateIanaTimezone } from '../types';
+
+/**
+ * IANA timezones used in the codebase
+ * Note - If needing a full list, we should look to Intl.supportedValuesOf('timeZone')
+ * {@link https://www.iana.org/time-zones}
+ */
+export const DATE_IANA_TIMEZONES = { LONDON: 'Europe/London' } as const satisfies Record<string, DateIanaTimezone>;
```
Comment (abhi-markan) [inline libs/common/src/constants/date-iana-timezones.ts:8]: Can this please be picked from the env variable `TZ`.

**Review by abhi-markan — CHANGES_REQUESTED** [review]

File: libs/common/src/constants/date-iana-timezones.ts (hunk 6 above)
Comment (AlexBramhill) [reply libs/common/src/constants/date-iana-timezones.ts:8]: I'm slightly cautious of this -- (prior to this pr) our code currently references everywhere as `LONDON_TIMEZONE` which is referencing the IANA timezone name, which is static so shouldn't be an env var. If we're after `DEFAULT_TIMEZONE` then I think we should look to refactor the variables, which is outside this scope (and will need to be considered whereever we transform timezones -- ie if we changes the `TZ` var to Europe/Paris, would we want to change _all_ cases of where we show time for this? Would we want to save things in the database like this etc?

Basically -- I know you'll have a great reason for this -- but for now as this is such a big ticket etc, and the changes could introduce subtle bugs if not considered, I might leave it if okay.

Comment (ttbarnes) [reply libs/common/src/constants/date-iana-timezones.ts:8]: If we want to use env vars instead of importing a constant, I suggest we do this in a different PR.

This PR is purely aimed at creating a DRY/common instance of `Europe/London`, some of which were hard coded.

Also btw @AlexBramhill maybe should rename the PR title to be `chore` instead of `feature`

Comment (AlexBramhill) [reply libs/common/src/constants/date-iana-timezones.ts:8]: updated title

Comment (abhi-markan) [reply libs/common/src/constants/date-iana-timezones.ts:8]: It is a minor change with values always set to `Europe/London` reason I suggest using `TZ` is to ensure all the Timezone values are referred from a single point during microservice initialisation locally, CICD or on Azure.

@ttbarnes understand this PR aims to make it DRY what I am suggesting is within those remits by not repeating a value which is already specified in env, existing PR or a new PR is both acceptable as long as it is updated since time is being invested in keeping the code DRY.

Comment (ttbarnes) [reply libs/common/src/constants/date-iana-timezones.ts:8]: Thanks @abhi-markan makes sense to me, I think we should do it as part of this PR whilst we're here. Unless it's blocking SSO in which case i'd suggest a new PR.

I understand Alex's concern about changing the TZ var to another timezone, but this seems like an extremely rare potential future requirement to me. **UK**EF

Comment (AlexBramhill) [reply libs/common/src/constants/date-iana-timezones.ts:8]: Could I ask why we need to use an env var for it please?

Comment (AlexBramhill) [reply libs/common/src/constants/date-iana-timezones.ts]: I'm not sure I agree this should be an env var. We do not have checking across services that these env vars are all identical, and its important to ensure they are all consistently set. Further, we have no way to verify that all our env vars are set to a valid IANA time zone. 

Regardless:
- I've updated this pattern to use a config, as it is no longer a constant.
- I've added validation that the env var is a string (so it'll fail if missing). 
- I've renamed the timezone to `DEFAULT` to reflect that this is the `DEFAULT` timezone so it won't be incorrectly used by people looking for `LONDON` explicitly (If we ever have to use the `TZ` env var, we will most likely need to reference LONDON timezone in the future, considering UKEF is a UK business)
- I've not added a default to this env var, as adding such a fallback would allow services to easily get out of sync accidentally
- I've had a brief skim of the code, and I don't think this will have an effect on existing code, but timezone changes can be prone to error.

Comment (AlexBramhill) [reply libs/common/src/constants/date-iana-timezones.ts]: I would like to pull the above change, and keep it static, as it adds additional complexity for the release of SSO which is out of scope (as seen in all tests failing due to env vars not being configured properly). Mind if I revert @abhi-markan

File: .github/workflows/development_infrastructure.yml [hunk 7]
```diff
@@ -1126,7 +1126,8 @@ jobs:
                 GOV_NOTIFY_API_KEY="${{ secrets.GOV_NOTIFY_API_KEY }}" \
                 MONGO_INITDB_DATABASE="${{ env.database_name }}" \
                 MONGODB_URI="$mongo_connection_string" \
-                RATE_LIMIT_THRESHOLD=${{ vars.RATE_LIMIT_THRESHOLD }}
+                RATE_LIMIT_THRESHOLD=${{ vars.RATE_LIMIT_THRESHOLD }} \
+                TZ="${{ vars.TZ }}"
```
Comment (abhi-markan) [inline .github/workflows/development_infrastructure.yml:1130]:
```suggestion
                TZ=${{ vars.TZ }}
```

**Review by abhi-markan — CHANGES_REQUESTED** [review]
One minor commit please.

**Review by abhi-markan — APPROVED** [review]
